### PR TITLE
Add support for DRSDT

### PR DESCRIPTION
**Pull request by totto82**
This is a rebased, squashed and cleaned version of #239 

@hnil Did I miss something? 
@andlaus Can you review this? 

Tested on SPE1CASE1, SPE1CASE2, SPE1CASEINF and SPE1SPEMID

TODO: test it on the bigmodels to make sure they doesn't regress. 

Merged together with OPM/opm-material#269

### CODE REVIEW

File: ewoms/models/blackoil/blackoilintensivequantities.hh [hunk 1]
```diff
@@ -219,7 +226,12 @@ public:
             // if the switching variable is the mole fraction of the gas component in the
             // oil phase, we can directly set the composition of the oil phase
             const auto& Rs = priVars.makeEvaluation(Indices::compositionSwitchIdx, timeIdx);
-            fluidState_.setRs(Rs);
+            //if(FluidSystem::enableRateLimitedDissolvedGas() == Opm::FluidSystems::All){
+            //    const auto& RsLimited = rateLimitedDissolvedGasUpdate(Rs, So, elemCtx,dofIdx,timeIdx);
+            //    fluidState_.setRs(RsLimited);
+            //} else {
+                fluidState_.setRs(Rs);
+            //}
```
Comment (totto82) [inline ewoms/models/blackoil/blackoilintensivequantities.hh]: @hnil Why was this commented out?

File: ebos/eclproblem.hh [hunk 2]
```diff
@@ -477,6 +477,9 @@ public:
             Scalar initialDt = EWOMS_GET_PARAM(TypeTag, Scalar, InitialTimeStepSize);
             dt = std::min(dt, initialDt);
         }
+        // The updateCell values is called before dt is updated to since it is based on the prevois step.
+        // this put the model state set values which should be from the previous step
```
Comment (andlaus) [inline ebos/eclproblem.hh]: I don't think that the model should be concerned with "cell values". for one thing what are the "celll values" that are supposed to be updated, for another, models are not specific for element-centered FV. Move the contents of this method to EclProblem, i.e. this call boils down to `updateElementValues_()`  (more descriptive name, please!) in `beginTimeStep()`

File: ewoms/models/blackoil/blackoilintensivequantities.hh [hunk 3]
```diff
@@ -199,7 +200,13 @@ public:
                                                             oilPhaseIdx,
                                                             pvtRegionIdx,
                                                             SoMax);
-                fluidState_.setRs(RsSat);
+                if(FluidSystem::enableRateLimitedDissolvedGas() > Opm::FluidSystems::None){
```
Comment (andlaus) [inline ewoms/models/blackoil/blackoilintensivequantities.hh]: I am quite sure that the fluid system should not need to care if the gas-in-oil or the oil-in-gas is rate limited or not, i.e. please move this flag to the model.

File: ewoms/models/blackoil/blackoilintensivequantities.hh [hunk 4]
```diff
@@ -199,7 +200,13 @@ public:
...
+                    // limit the updated Rs value by a given rate pr volume
+                    const Evaluation& RsLimited = rateLimitedDissolvedGasUpdate(RsSat, So, elemCtx, dofIdx, timeIdx);
```
Comment (andlaus) [inline ewoms/models/blackoil/blackoilintensivequantities.hh]: I don't think that the `IntensiveQuantities` need to concern itself with the rate limits. the only place where this is needed is the update of the primary variables.

File: ewoms/models/blackoil/blackoilintensivequantities.hh (hunk 1 above)
Comment (andlaus) [reply ewoms/models/blackoil/blackoilintensivequantities.hh]: I think the old version is correct, as the IQs should not need to consider rate limits. that said, please remove commented out code instead of leaving it in place (commented out code just clutters the the files and makes them less maintainable. IMO)

File: ewoms/models/blackoil/blackoilintensivequantities.hh [hunk 5]
```diff
@@ -344,6 +356,18 @@ public:
 #endif
     }
 
+    /*! function to update Rs with rate limmitation this is used if keyword DRSDT is present in configuration file
+     */
+    Evaluation rateLimitedDissolvedGasUpdate(const Evaluation& RsE,const Evaluation& So, const ElementContext& elemCtx, unsigned dofIdx, unsigned timeIdx)
```
Comment (andlaus) [inline ewoms/models/blackoil/blackoilintensivequantities.hh]: the IQs do not concern themselfs with updating the primary variables, so this method probably does not belong here.

File: ewoms/models/blackoil/blackoilmodel.hh [hunk 6]
```diff
@@ -220,7 +223,12 @@ public:
     BlackOilModel(Simulator& simulator)
         : ParentType(simulator)
     {}
-
+    enum CellValuesName : unsigned {
+        soPrev = 0,
+        rsPrev = 1,
+        soMax = 2,
+        numCellValues,
+    };
```
Comment (andlaus) [inline ewoms/models/blackoil/blackoilmodel.hh]: as @totto82 already mentioned in a comment related to these PRs: changing the approach of how to access the data of the model is out of scope of this PR and should thus be discussed separately. Please remove (for now)

File: ewoms/models/blackoil/blackoilmodel.hh [hunk 7]
```diff
@@ -241,7 +249,11 @@ public:
      */
     void finishInit()
     {
-        maxOilSaturation_.resize(this->numGridDof(), 0.0);
+        cellValues_.resize(numCellValues);
+        cellValues_.resize(this->numGridDof());
+        for(unsigned i=0; i < cellValues_.size() ; ++i){
+            cellValues_[i].fill( 0.0);
+        }
```
Comment (andlaus) [inline ewoms/models/blackoil/blackoilmodel.hh]: "cell values" stuff -> please move to a separate PR

File: ewoms/models/blackoil/blackoilmodel.hh [hunk 8]
```diff
@@ -384,8 +396,13 @@ public:
         outstream << priVars.primaryVarsMeaning() << " ";
         outstream << priVars.pvtRegionIndex() << " ";
 
-        if (maxOilSaturation_.size() > 0)
-            outstream << maxOilSaturation_[dofIdx] << " ";
+        if (cellValues_.size()>0){
```
Comment (andlaus) [inline ewoms/models/blackoil/blackoilmodel.hh]: same

File: ewoms/models/blackoil/blackoilmodel.hh [hunk 9]
```diff
@@ -422,8 +439,13 @@ public:
         unsigned pvtRegionIdx;
         instream >> pvtRegionIdx;
 
-        if (maxOilSaturation_.size() > 0)
-            instream >> maxOilSaturation_[dofIdx];
+        if (cellValues_.size()>0){
```
Comment (andlaus) [inline ewoms/models/blackoil/blackoilmodel.hh]: ditto

File: ewoms/models/blackoil/blackoilmodel.hh [hunk 10]
```diff
@@ -477,14 +499,26 @@ public:
      * match the results of the 'flow' and ECLIPSE 100 simulators.
      */
     Scalar maxOilSaturation(unsigned globalDofIdx) const
```
Comment (andlaus) [inline ewoms/models/blackoil/blackoilmodel.hh]: as discussed earlier, this function should not be here. could you please move it to `EclProblem` and add a dummy function to `BlackOilProblem` that just returns 1?

File: ewoms/models/blackoil/blackoilmodel.hh [hunk 11]
```diff
@@ -477,14 +499,26 @@ public:
...
+    { cellValues_[globalDofIdx][soMax] = value; }
+
+    Scalar cellValues(unsigned globalDofIdx,CellValuesName cv) const
+    { return cellValues_[globalDofIdx][cv]; }
+
+    /*!
+     * \brief Sets an elements maximum oil phase saturation observed during the
+     *        simulation (used for restarting from UNRST-files).
+     */
+    void setCellValue(const Scalar& value, unsigned globalDofIdx, CellValuesName cv)
```
Comment (andlaus) [inline ewoms/models/blackoil/blackoilmodel.hh]: "cell values" stuff

File: ewoms/models/blackoil/blackoilmodel.hh [hunk 12]
```diff
@@ -494,34 +528,25 @@ public:
      * simulation it sometimes needs to be called after a time step or before an episode
      * starts.
      */
-    void updateMaxOilSaturations()
+    void updateCellValues()
```
Comment (andlaus) [inline ewoms/models/blackoil/blackoilmodel.hh]: updating quasi-static data after a time step is the domain of the problem: this method does not make any sense for e.g. vertex-centered FV discretizations

File: ewoms/models/blackoil/blackoilprimaryvariables.hh [hunk 13]
```diff
@@ -381,14 +382,21 @@ public:
                 Scalar po = (*this)[Indices::pressureSwitchIdx];
                 Scalar T = asImp_().temperature_();
                 Scalar SoMax = problem.model().maxOilSaturation(globalDofIdx);
-                Scalar RsSat = FluidSystem::oilPvt().saturatedGasDissolutionFactor(pvtRegionIdx_,
+                Scalar Rs = FluidSystem::oilPvt().saturatedGasDissolutionFactor(pvtRegionIdx_,
                                                                                    T,
                                                                                    po,
                                                                                    So2,
                                                                                    SoMax);
```
Comment (andlaus) [inline ewoms/models/blackoil/blackoilprimaryvariables.hh]: probably, it is better to add RsMax and RsMin and use

```
Scalar RsMax = std::min(RsSat, RsOld + RsMaxDelta);
Scalar RsMin = std::max(0.0, RsOld - RsMaxDelta);
```

where `RsOld` and `RsMaxDelta` are retrieved from the problem like `SoMax`.

**Comment by andlaus** [comment]
I have a bit more general question related to DRSDT: how should composition changes due to pressure changes in the saturated zone be handled? e.g., pressure increases in an area where all three phases are present and the increase of the saturated Rs increases the value set by DRSDT.

File: ewoms/models/blackoil/blackoilintensivequantities.hh (hunk 1 above)
Comment (totto82) [reply ewoms/models/blackoil/blackoilintensivequantities.hh]: I just wasn't sure why the commented out code was added in the first place.

**Comment by andlaus** [comment]
Ah, I forgot: With this, the `UgGridHelpers` stuff can finally be replaced in that code by direct usage of the DUNE grid interface without breaking anything. this would enable alternative grids like `Dune::PolyhedralGrid` or `Dune::ALUGrid` again. Unfortunately, it is quite a bit of non-trivial work.
